flow/Platform.cpp

```diff
@@ -1843,15 +1889,22 @@ std::string basename( std::string const& filename ) {
 }
 
 std::string parentDirectory( std::string const& filename ) {
-	auto abs = abspath(filename);
-	size_t sep = abs.find_last_of( CANONICAL_PATH_SEPARATOR );
-	if (sep == std::string::npos) {
-		TraceEvent(SevError, "GetParentDirectoryOfFile")
+	std::string abs = abspath(filename);
+
+	// Can't take the parent of empty, root, or anything ending in an unresolved ..
+	if(abs.empty()
+		|| (abs.size() == 1 && abs.front() == CANONICAL_PATH_SEPARATOR)
+		|| abs == ".."
```

Comment:
Here the path `abs` should have already been canonicalized by `abspath()` where either the system call or the `cleanPath` function already does the job.

Reply:
These checks are actually meant to handle the possible results of cleanPath(), not absPath().  I changed it at some point and didn't update the checks, however I actually now believe it should be changed back.  That is to say:  parentDirectory should not consult the filesystem, and hence not resolve symbolic links.  This is the behavior that fdbmonitor needs from parentDirectory() because it wants the monitor the configuration file for changes, including if the directory it was read from was actually a symbolic link which was then modified to point to a different directory.

This also makes parentDirectory trivial to implement using cleanPath(), as parentDirectory(x) = cleanPath(joinPath(x, "..")).  And if the caller wants to resolve that to an absolute path it can call absPath().